src/storage/src/controller.rs

```diff
@@ -1496,44 +1500,65 @@ mod persist_read_handles {
+                                                    // to empty antichain (aka we dropped it).
+                                                    Err(StorageError::IdentifierMissing(id))
+                                                }
+                                            };
+                                            oneshot.send(result).expect("Oneshot should not fail");
                                         }
                                         PersistWorkerCmd::Shutdown => {
                                             shutdown = true;
                                         }
                                     }
                                 }
 
-
+                                let mut drops = HashSet::new();
                                 if !downgrades.is_empty() {
                                     let futs = FuturesUnordered::new();
 
                                     for (id, read) in read_handles.iter_mut() {
                                         if let Some((span, since)) = downgrades.remove(id) {
-                                            let fut = async move {
-                                                // Use maybe_downgrade_since here so that we opt
-                                                // into rate-limiting. It's okay for the since to
-                                                // lag behind a bit and this _greatly_ reduces the
-                                                // persist traffic.
-                                                read.maybe_downgrade_since(&since).instrument(span).await;
-                                            };
+                                        // A None read handle is one that had a successful
+                                        // downgrade_since to the empty antichain, so we treat it as
+                                        // a no-op for any later downgrade_since calls.
+                                        if let Some(read) = read.as_mut() {
+                                                // If we downgrade_since to the empty antichain,
+                                                // then expire and drop the ReadHandle afterward.
+                                                if since.is_empty() {
+                                                    drops.insert(*id);
+                                                }
+
+                                                let fut = async move {
+                                                    // Use maybe_downgrade_since here so that we opt
+                                                    // into rate-limiting. It's okay for the since to
+                                                    // lag behind a bit and this _greatly_ reduces the
+                                                    // persist traffic.
+                                                    read.maybe_downgrade_since(&since).instrument(span).await;
```

Comment:
Should this be a `maybe_downgrade_since` when we downgrade to the empty frontier? It seems we would want to make sure that a downgrade to the empty frontier goes through, because we will not supersede that with any later calls.

Reply:
good eye! I think this is fine in the short-term because missing the downgrade only means persist doesn't get to clean up as much, but we're not doing that (yet) anyway. a patch for this seems pretty easy, but I think maybe I just want to wait for the ReadHandle change? I actually don't think it's so hard (even compared to your since.is_empty() idea), I'd bet most of the time will be writing tests for it

Reply:
There is a subtle difference: All the downgrade logic only downgrades the "global" since (which only exists morally, as the meet of the reader-local sinces) when there is at least one reader. If at least one reader downgrades to `[]` and then all readers leave, the shard global since will be `[]`. If all the readers just `expire()` themselves, the global since will not be downgraded.

But now that I type it I'm not 100% sure whether this is still the current behaviour.

Reply:
yep, putting a `downgrade_since([])` with the drops would definitely work

for better or for worse, `expire` works a little differently than `downgrade_since([])` on purpose (though happy to discuss). imagine that environmentd is down (for whatever reason, maybe it crashed and aws is slow giving us a new machine) and the only active read lease is in some computed, we don't (or I think we don't) want an `expire` call in `computed` to render the shard permanently unreadable. so, persist at the moment, will only move the since forward in response to explicit `downgrade_since` calls.

there's still the possibility that the computed would call `downgrade_since([])` and hose the shard, but I'm not sure what to do about that. we're okay if environmentd comes back before the old one's read lease expires (15m currently, but that was selected arbitrarily). probably the only thing saving us at the moment is that persist_source doesn't downgrade the since past the upper

Reply:
ha! aljoscha beat me to it